Provide NVIDIA CUDA build data in metadata and API

angerson opened the pull request:
This change:

First exposes //third_party/gpus:find_cuda_config as a library.

Then, it extends gen_build_info.py with find_cuda_config to provide package
build information within TensorFlow's API. This is accessible as a
dictionary:

    from tensorflow.python.platform import build_info
    print(build_info.cuda_build_info)
    {'cuda_version': '10.2', 'cudnn_version': '7', 'tensorrt_version': None, 'nccl_version': None}

Finally, setup.py pulls that into package metadata (I also formatted setup.py). The same wheel's
long description ends with:

    TensorFlow 2.1.0 for NVIDIA GPUs was built with these platform
    and library versions:

      - NVIDIA CUDA 10.2
      - NVIDIA cuDNN 7
      - NVIDIA NCCL not enabled
      - NVIDIA TensorRT not enabled

In lieu of [NVIDIA CUDA classifiers][1], the same metadata is exposed in the
normally-unused "platform" tag:

    >>> import pkginfo
    >>> a = pkginfo.Wheel('./tf_nightly_gpu-2.1.0-cp36-cp36m-linux_x86_64.whl')
    >>> a.platforms
    ['cuda_version:10.2', 'cudnn_version:7', 'tensorrt_version:None', 'nccl_version:None']

I'm not 100% confident this is the best way to accomplish this. It
seems odd to import like this setup.py, even though it works, even in
an environment with TensorFlow installed.

One caveat for RBE: the contents of genrules still run on the local
system, so I had to syncronize my local environment with the RBE
environment I used to build TensorFlow. I'm not sure if this is going to
require intervention on TensorFlow's current CI.

Currently tested only on Linux GPU (Remote Build) for Python 3.6. I'd
like to see more tests before merging.

Resolves https://github.com/tensorflow/tensorflow/issues/38351.

[1]: https://github.com/pypa/trove-classifiers/issues/25

angerson committed:
Provide NVIDIA CUDA build data in metadata and API

This change:

First exposes //third_party/gpus:find_cuda_config as a library.

Then, it extends gen_build_info.py with find_cuda_config to provide package
build information within TensorFlow's API. This is accessible as a
dictionary:

    from tensorflow.python.platform import build_info
    print(build_info.cuda_build_info)
    {'cuda_version': '10.2', 'cudnn_version': '7', 'tensorrt_version': None, 'nccl_version': None}

Finally, setup.py pulls that into package metadata. The same wheel's
long description ends with:

    TensorFlow 2.1.0 for NVIDIA GPUs was built with these platform
    and library versions:

      - NVIDIA CUDA 10.2
      - NVIDIA cuDNN 7
      - NVIDIA NCCL not enabled
      - NVIDIA TensorRT not enabled

In lieu of NVIDIA CUDA classifiers [1], the same metadata is exposed in the
normally-unused "platform" tag:

    >>> import pkginfo
    >>> a = pkginfo.Wheel('./tf_nightly_gpu-2.1.0-cp36-cp36m-linux_x86_64.whl')
    >>> a.platforms
    ['cuda_version:10.2', 'cudnn_version:7', 'tensorrt_version:None', 'nccl_version:None']

I'm not 100% confident this is the best way to accomplish this. It
seems odd to import like this setup.py, even though it works, even in
an environment with TensorFlow installed.

One caveat for RBE: the contents of genrules still run on the local
system, so I had to syncronize my local environment with the RBE
environment I used to build TensorFlow. I'm not sure if this is going to
require intervention on TensorFlow's current CI.

Currently tested only on Linux GPU (Remote Build) for Python 3.6. I'd
like to see more tests before merging.

[1]: (https://github.com/pypa/trove-classifiers/issues/25),

angerson commented:
Tagging parties I know are interested: @gabrieldemarmiesse, @gunan, @sub-mod, @perfinion

gunan commented:
@alextp for API review

alextp commented:
@gunan I don't see any API changes

gunan commented:
@angerson, are you planning to add these to tf.config in a later change?

angerson committed:
Add NVIDIA CUDA and cuDNN info to tf.config

angerson commented:
Ah, I hadn't fully understood what makes something part of the public API. I've added `tf.config.get_cuda_version_used_to_compile_tf()` and `tf.config.get_cudnn_version_used_to_compile_tf()`.

gabrieldemarmiesse commented:
You don't have to take exactly the names that I proposed, it was just an example :)

alextp commented on tensorflow/python/framework/config.py:
```diff
@@ -653,3 +654,25 @@ def disable_mlir_bridge():
+
+
+@tf_export('config.get_cuda_version_used_to_compile_tf')
+def get_cuda_version_used_to_compile_tf():
```
Why not a config.get_build_info that returns a dictionary with string keys representing all relevant build info (including cuda and cudnn versions)?

byronyi replied:
Agreed. A dictionary would be great so we do not need to add `get_rocm_version_used_to_compile_tf` or `get_hccl_version_used_to_compile_tf` etc. in the future.

angerson replied:
Agreed, I've addressed this in my latest commit.

angerson committed:
Convert build_info to dict format and expose it.

Since this module now generates a dictionary to expose in tf.config, it
doesn't make much sense to store only certain values in the build_info
dictionary and others as module variables. This obsoletes a lot of code
in gen_build_info.py and I've removed it.

I also updated all the in-code references I've found to the build_info
module. I think this may break whomever used to be using the build_info
library, but since it wasn't part of the API, there was no guarantee
that it would continue to be available.

angerson commented:
Thanks for your feedback! I replaced the functions with `tf.config.get_build_config()`, which returns a dictionary. I had planned to use a NamedTuple instead, but I feel that a Tuple implies more stability than a dictionary; the build info exposed here is not guaranteed to contain all of these fields, and a dictionary has a standard for using that (`dict.get()`).

Also, storing all of the information in a dict seems to remove the need to store any other module-level variables in the generated python file, so I've moved everything into the dict as key/value pairs, including the is_config_xxx flags, which don't seem to warrant needing specific flags.

angerson committed:
Fix incorrect call to tf_export

angerson committed:
Fix typos from the latest version

gbaned commented:
@angerson Can you please fix build failures ? Thanks!

angerson committed:
Remove extra debug print statement

sub-mod commented:
@angerson is it possible to backport this to 1.15.x branch ?

mihaimaruseac commented:
@sub-mod Sure, we can backport and merge when ready to do a patch release

gunan commented:
Unfortunately, no.
All past release branches are closed for new features, even 2.2

mihaimaruseac commented:
Oh, Gunhan is right, this is a new feature, not a bug fix

perfinion commented:
@angerson @alextp @gunan These new API calls should be under `tf.sysconfig.foo` not under `tf.config.foo`

The other build-time type things are already under tf.sysconfig eg `tf.sysconfig.get_include()` or `tf.sysconfig.get_compile_flags()`

Theres also `tf.test.is_built_with_cuda()` I think those should move under `tf.sysconfig` too but they are already released so they'd have to stay as aliases.

alextp commented:
I don't thinkin tf.sysconfig is a public namespace.

On Tue, May 5, 2020 at 11:06 AM Jason Zaman <notifications@github.com>
wrote:

> @angerson <https://github.com/angerson> @alextp
> <https://github.com/alextp> @gunan <https://github.com/gunan> These new
> API calls should be under tf.sysconfig.foo not under tf.config.foo
>
> The other build-time type things are already under tf.sysconfig eg
> tf.sysconfig.get_include() or tf.sysconfig.get_compile_flags()
>
> Theres also tf.test.is_built_with_cuda() I think those should move under
> tf.sysconfig too but they are already released so they'd have to stay as
> aliases.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/38964#issuecomment-624216895>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAABHRMENSKBS5WPSV2PKJLRQBIRPANCNFSM4MSLY43A>
> .
>


-- 
 - Alex

alextp commented:
Oh nevermind it is a public namespace.

On Tue, May 5, 2020 at 11:27 AM Alexandre Passos <apassos@google.com> wrote:

> I don't thinkin tf.sysconfig is a public namespace.
>
> On Tue, May 5, 2020 at 11:06 AM Jason Zaman <notifications@github.com>
> wrote:
>
>> @angerson <https://github.com/angerson> @alextp
>> <https://github.com/alextp> @gunan <https://github.com/gunan> These new
>> API calls should be under tf.sysconfig.foo not under tf.config.foo
>>
>> The other build-time type things are already under tf.sysconfig eg
>> tf.sysconfig.get_include() or tf.sysconfig.get_compile_flags()
>>
>> Theres also tf.test.is_built_with_cuda() I think those should move under
>> tf.sysconfig too but they are already released so they'd have to stay as
>> aliases.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/tensorflow/tensorflow/pull/38964#issuecomment-624216895>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAABHRMENSKBS5WPSV2PKJLRQBIRPANCNFSM4MSLY43A>
>> .
>>
>
>
> --
>  - Alex
>


-- 
 - Alex

perfinion commented:
https://www.tensorflow.org/api_docs/python/tf/sysconfig I've never heard its not public? Its been in the docs for ages and theres no indication its not to be relied on? There isnt much in it so its likely overlooked most of the time but its definitely listed publicly.

I know for sure custom-op uses it to make the build rules so it cant be removed either.

angerson commented:
It's not reflected here in this PR, but I've moved the call into tf.sysconfig in the being-internally-reviewed change.

r4nt commented:
I'm proposing a rollback internally - I think the right way to do this is to fill in a template file from the repo rule (cuda_configure.bzl) instead of trying to call into find_cuda_config as part of the action, which as a genrule is marked "local" and can thus run on a different machine, which is brittle.

r4nt commented on tensorflow/tools/build_info/gen_build_info.py:
```diff
@@ -19,50 +19,58 @@
+    # before calling find_cuda_config.
+    backup_argv = sys.argv
+    sys.argv = libs
+    cuda = find_cuda_config.find_cuda_config()
```
Basically, instead of calling find_cuda_config here, I'd create library cuda_config in cuda_configure.py that contains the exact same entries we've configured the build with. We can then depend on that py_library here, import it, and read the values as recorded during the repo configuration.

r4nt commented:
Just noticed that this rule also has no dependency on running locally, so I think just making it a normal genrule and removing local=1 will work fine.

That said, I still think that we should not compute this info twice, as that seems more brittle than it needs to be.
